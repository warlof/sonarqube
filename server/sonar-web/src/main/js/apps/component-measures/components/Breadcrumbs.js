/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
// @flow
import React from 'react';
import Breadcrumb from './Breadcrumb';
import { getBreadcrumbs } from '../../../api/components';
import type { Component } from '../types';

type Props = {|
  className?: string,
  component: Component,
  handleSelect: Component => void,
  rootComponent: Component
|};

type State = {
  breadcrumbs: Array<Component>
};

export default class Breadcrumbs extends React.PureComponent {
  mounted: boolean;
  props: Props;
  state: State = {
    breadcrumbs: []
  };

  componentDidMount() {
    this.mounted = true;
    this.fetchBreadcrumbs(this.props);
  }

  componentWillReceiveProps(nextProps: Props) {
    if (this.props.component !== nextProps.component) {
      this.fetchBreadcrumbs(nextProps);
    }
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  fetchBreadcrumbs = ({ component, rootComponent }: Props) => {
    const isRoot = component.key === rootComponent.key;
    if (isRoot) {
      if (this.mounted) {
        this.setState({ breadcrumbs: isRoot ? [component] : [rootComponent, component] });
      }
      return;
    }
    getBreadcrumbs(component.key).then(breadcrumbs => {
      if (this.mounted) {
        this.setState({ breadcrumbs });
      }
    });
  };

  render() {
    const { breadcrumbs } = this.state;
    if (breadcrumbs.length <= 0) {
      return null;
    }
    const lastItem = breadcrumbs[breadcrumbs.length - 1];
    return (
      <div className={this.props.className}>
        {breadcrumbs.map(component =>
          <Breadcrumb
            key={component.key}
            canBrowse={component.key !== lastItem.key}
            component={component}
            isLast={component.key === lastItem.key}
            handleSelect={this.props.handleSelect}
          />
        )}
      </div>
    );
  }
}
