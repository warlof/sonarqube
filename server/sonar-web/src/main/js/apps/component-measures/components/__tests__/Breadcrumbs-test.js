/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import React from 'react';
import { mount } from 'enzyme';
import Breadcrumbs from '../Breadcrumbs';
import { doAsync } from '../../../../helpers/testUtils';

jest.mock('../../../../api/components', () => ({
  getBreadcrumbs: () =>
    Promise.resolve([
      { key: 'anc1', name: 'Ancestor1' },
      { key: 'anc2', name: 'Ancestor2' },
      { key: 'bar', name: 'Bar' }
    ])
}));

it('should display correctly for the list view', () => {
  const wrapper = mount(
    <Breadcrumbs
      component={{ key: 'bar', name: 'Bar' }}
      handleSelect={() => {}}
      rootComponent={{ key: 'foo', name: 'Foo' }}
      view="list"
    />
  );
  expect(wrapper).toMatchSnapshot();
});

it('should display only the root component', () => {
  const wrapper = mount(
    <Breadcrumbs
      component={{ key: 'foo', name: 'Foo' }}
      handleSelect={() => {}}
      rootComponent={{ key: 'foo', name: 'Foo' }}
      view="tree"
    />
  );
  expect(wrapper.state()).toMatchSnapshot();
});

it.only('should load the breadcrumb from the api', () => {
  const wrapper = mount(
    <Breadcrumbs
      component={{ key: 'bar', name: 'Bar' }}
      handleSelect={() => {}}
      rootComponent={{ key: 'foo', name: 'Foo' }}
      view="tree"
    />
  );
  return doAsync(() => {
    expect(wrapper.state()).toMatchSnapshot();
  });
});
