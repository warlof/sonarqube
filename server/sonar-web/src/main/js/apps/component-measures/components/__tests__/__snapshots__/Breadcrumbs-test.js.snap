// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display correctly for the list view 1`] = `
<Breadcrumbs
  component={
    Object {
      "key": "bar",
      "name": "Bar",
    }
  }
  handleSelect={[Function]}
  rootComponent={
    Object {
      "key": "foo",
      "name": "Foo",
    }
  }
  view="list"
>
  <div>
    <Breadcrumb
      canBrowse={true}
      component={
        Object {
          "key": "foo",
          "name": "Foo",
        }
      }
      handleSelect={[Function]}
      isLast={false}
    >
      <span>
        <a
          href="#"
          onClick={[Function]}
        >
          Foo
        </a>
        <span
          className="slash-separator"
        />
      </span>
    </Breadcrumb>
    <Breadcrumb
      canBrowse={false}
      component={
        Object {
          "key": "bar",
          "name": "Bar",
        }
      }
      handleSelect={[Function]}
      isLast={true}
    >
      <span>
        <span>
          Bar
        </span>
      </span>
    </Breadcrumb>
  </div>
</Breadcrumbs>
`;

exports[`should display only the root component 1`] = `
Object {
  "breadcrumbs": Array [
    Object {
      "key": "foo",
      "name": "Foo",
    },
  ],
}
`;

exports[`should load the breadcrumb from the api 1`] = `
Object {
  "breadcrumbs": Array [
    Object {
      "key": "anc1",
      "name": "Ancestor1",
    },
    Object {
      "key": "anc2",
      "name": "Ancestor2",
    },
    Object {
      "key": "bar",
      "name": "Bar",
    },
  ],
}
`;
